### PR TITLE
[ID-98] Rate limits

### LINKED ISSUE #98: [FEATURE] Rate limits
**Is your feature request related to a problem? Please describe.**
Currently, all notifications are delivered as rapidly as possible to all specified recipients. This has resulted in extremely large spikes in system usage in cases where notifications are sent to the large groups (eg. an entire school). Usage spikes like these are not handled well by autoscalers and must be avoided. 

**Describe the solution you'd like**
We should develop a rate-limiting system to ensure notifications are not sent all at once to too many recipients. This system should batch notifications into manageable chunks and wait some appropriate amount of time between sending notifications to each one. These notification batches should be entered into a sorted queue. We can then schedule a timer for the first item in the queue to be sent. This queue should be stored in the database along with a lock to prevent multiple instances of the Notifications BB from sending the same message. All relevant thresholds should be dynamically configurable by admins.

**Describe alternatives you've considered**
An alternative would be to randomly assign each user to a different group when they register with the Notifications BB. We would need to enforce a maximum size on these user groups that should again be configurable. When a notification is sent, rather than batching the notifications, we would send the notification to one user group at a time.

### PR DESCRIPTION
**Pull request by petyos**
## Description

**Resolves #98 

Here are the highlihts of this PR:
- Introduced priority queue feature for sending notifications.
- You can send notification immediatelly or in a schedulled time in the future.
- If you want to send a message at specific time in the future then you just need to add "time" field when hitting the API for creating messages. If the "time" field is omitted then it means now and tries to send the notification asap.
- You can also set a priority of this message from 1 to 10. 1 means highest priority.
- Here is the priority order for sending: the first criteria is the "time" field. The notifications with most recently time will be sent first. If we have notifications with equals time then the second criteria is the "priority" field mentioned above.
- The queue should manage a lot of notifications processing them on small pieces. The piece count is managed from the database. 50 is acceptable number.
- Once one service starts processing the priority queue then it locks it until it finish. Once finished then it unlocks it again. This way we prevent from multiple services sending notifications simulativelly.
- Removed code for data manipualting for multi-tenancy and recipients refactoring - no needed anymore.
- Code order in the web driver adapter.
- Removed mode.InputMessage and introduced tool generated struct on web driver level for the create messages APIs.

[comment]: # (This project only accepts pull requests related to open issues.)
[comment]: # (If suggesting a new feature or change, please discuss it in an issue first.)
[comment]: # (If fixing a bug, there should be an issue describing it with steps to reproduce.)

## Review Time Estimate
_Please give your idea of how soon this pull request needs to be reviewed by selecting one of the options below. This can be based on the criticality of the issue at hand and/or other relevant factors._

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Immediately
- [ ] Within a week
- [x] When possible

## Type of changes
_Please select a relevant option:_

[comment]: # (To select an option, please put an 'x' in the applicable box.)
[comment]: # (If you're unsure about any of these, don't hesitate to ask. We're here to help!.)

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Other (any another change that does not fall in one of the above categories.)

### CODE REVIEW

**Comment by petyos** [comment]
When you test the changes you first need to:
- create collection "queue" in your database
- import the following record in it:
```
{
  "_id": "1",
  "status": "ready",
  "process_items_count": 50
}
```

**Review by mdryankov — APPROVED** [review]
LGTM

File: core/app_queue.go [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+	//timer
+	queueTimer *time.Timer
+	timerDone  chan bool
+}
+
+func (q queueLogic) start() {
+	q.logger.Info("queueLogic start")
+
+	q.processQueue()
```
Comment (shurwit) [inline core/app_queue.go:40]: I think there may be an issue here if the queue is locked when the instance first starts. In this case, it looks to me like the instance will never attempt to process the queue unless the specific instance receives a new message through the API. In normal functioning I think that is ok, but if some of the instances crash I think we could end up in a state where no instances are actively processing the timed items in the queue. One potential solution could be to periodically retry processing the queue until it is able to acquire the lock.

**Review by shurwit — APPROVED** [review]
Thanks @petyos, this is really great work! I think this implementation is really well structured and easy to follow. I did leave a couple of comments below talking about some edge cases that may come up if a thread or instance crashes. I don't believe that either one is critically urgent enough to stop this from being merged, but let me know if you have any questions or thoughts on them. Thanks again!

**Comment by petyos** [comment]
Hi @mdryankov , @shurwit , thanks for your feedback. I am merging it. @shurwit I've opened a separate task for covering your feedback here - https://github.com/rokwire/notifications-building-block/issues/140

**Pull request merged**